src/test/index.ts

```diff
@@ -71,103 +76,135 @@ export const test = _baseTest.extend<PlaywrightTestArgs & PlaywrightTestOptions,
+    await context.tracing.start({ name, screenshots: true, snapshots: true });
+  }
+
+  await use(context);
+
+  const testFailed = testInfo.status !== testInfo.expectedStatus;
+
+  const saveTrace = trace === 'on' || (testFailed && trace === 'retain-on-failure') || (trace === 'retry-with-trace' && testInfo.retry);
+  if (saveTrace) {
+    const tracePath = testInfo.outputPath(`trace.zip`);
+    await context.tracing.stop({ path: tracePath });
+  } else if (collectingTrace) {
+    await context.tracing.stop();
+  }
+
+  if (screenshot === 'on' || (screenshot === 'only-on-failure' && testFailed)) {
+    await Promise.all(allPages.map((page, index) => {
+      const screenshotPath = testInfo.outputPath(`test-${testFailed ? 'failed' : 'finished'}-${++index}.png`);
+      return page.screenshot({ timeout: 5000, path: screenshotPath }).catch(e => {});
+    }));
+  }
+  await context.close();
+
+  const deleteVideos = video === 'retain-on-failure' && !testFailed;
+  if (deleteVideos) {
+    await Promise.all(allPages.map(async page => {
+      const video = page.video();
+      if (!video)
+        return;
+      try {
+        const videoPath = await video.path();
+        await fs.promises.unlink(videoPath);
+      } catch (e) {
+        // Silent catch.
+      }
+    }));
+  }
+}
+
+async function withDebuggingContext(browser: Browser, params: PlaywrightTestOptions, testInfo: TestInfo, use: (context: BrowserContext) => Promise<void>) {
+  testInfo.setTimeout(0);
+  const options = mergeContextOptions(params);
+
+  if (!debuggingContext) {
+    const options = mergeContextOptions(params);
+    debuggingContextOptions = options;
+    debuggingContext = await browser.newContext(options);
+  } else {
+    expect(options).toEqual(debuggingContextOptions);
+  }
+
+  await debuggingContext.clearCookies();
+  await debuggingContext.clearPermissions();
+  await use(debuggingContext);
+}
+
+function mergeContextOptions(testOptions: PlaywrightTestOptions): BrowserContextOptions {
+  const topLevelContextOptions: BrowserContextOptions = {
+    acceptDownloads: testOptions.acceptDownloads,
+    bypassCSP: testOptions.bypassCSP,
+    colorScheme: testOptions.colorScheme,
+    deviceScaleFactor: testOptions.deviceScaleFactor,
+    extraHTTPHeaders: testOptions.extraHTTPHeaders,
+    geolocation: testOptions.geolocation,
+    hasTouch: testOptions.hasTouch,
+    httpCredentials: testOptions.httpCredentials,
+    ignoreHTTPSErrors: testOptions.ignoreHTTPSErrors,
+    isMobile: testOptions.isMobile,
+    javaScriptEnabled: testOptions.javaScriptEnabled,
+    locale: testOptions.locale,
+    offline: testOptions.offline,
+    permissions: testOptions.permissions,
+    proxy: testOptions.proxy,
+    storageState: testOptions.storageState,
+    timezoneId: testOptions.timezoneId,
+    userAgent: testOptions.userAgent,
+    viewport: testOptions.viewport,
+  };
+  return { ...testOptions.contextOptions, ...topLevelContextOptions };
```

Comment:
I think that if I do `test.use({ viewport: undefined })` to revert to the default, it will actually be undefined instead of the default. There is a helper against that: https://github.com/microsoft/playwright/blob/master/src/test/util.ts#L171-L180.